Source/Atomic/Network/MasterServerClient.cpp

```diff
@@ -0,0 +1,556 @@
+
+    unsigned char* ip = localEndPoint.ip;
+
+    char str[256];
+    sprintf(str, "%d.%d.%d.%d", (unsigned int)ip[0], (unsigned int)ip[1], (unsigned int)ip[2], (unsigned int)ip[3]);
+
+    Atomic::Network* network = GetSubsystem<Network>();
+    unsigned int localPort = network->GetServerPort();
+
+    String msg = String("{") +
+                 String("\"cmd\":") + String("\"registerServer\",") +
+                 String("\"internalIP\": \"") + String(str) + String("\", ") +
+                 String("\"internalPort\": ") + String(localPort) + String(", ") +
+                 String("\"id\":\"") + masterServerConnectionId_ + String("\", ") +
+                 String("\"serverName\":\"") + name + String("\"") +
+                 String("}");
+
+    SendMessageToMasterServer(msg);
+}
+
+void MasterServerClient::SendMessageToMasterServer(const String& msg)
+{
+    if (masterTCPConnection_)
+    {
+        String netString = String(msg.Length()) + ':' + msg;
+        masterTCPConnection_->Send(netString.CString(), netString.Length());
+    }
+    else
+    {
+        LOGERROR("No master server connection. Cannot send message");
+    }
+}
+
+void MasterServerClient::Update(float dt)
+{
+    if (masterTCPConnection_==NULL)
+    {
+        return;
+    }
+
+    CheckForMessageFromMaster();
+    ConnectToMasterUpdate(dt);
+    ConnectToGameServerUpdate(dt);
+    CheckForNatPunchThroughRequests(dt);
+}
+
+void MasterServerClient::CheckForNatPunchThroughRequests(float dt)
+{
+    if (clientIdToPunchThroughSocketMap_.Size() > 0)
+    {
+        if (timeTillNextPunchThroughAttempt_ <= 0)
+        {
+            for (HashMap<String, kNet::Socket*>::Iterator i = clientIdToPunchThroughSocketMap_.Begin();
+                 i != clientIdToPunchThroughSocketMap_.End();)
+            {
+                Network* network = GetSubsystem<Network>();
+                LOGINFO("Sending packet to client");
+                kNet::Socket* s = i->second_;
+
+                if (network->IsEndPointConnected(s->RemoteEndPoint()))
+                {
+                    i = clientIdToPunchThroughSocketMap_.Erase(i);
+                }
+                else
+                {
+                    s->Send("K",1);
+                    ++i;
+                }
+            }
+
+            // Reset the timer
+            timeTillNextPunchThroughAttempt_ = timeBetweenClientPunchThroughAttempts_;
+        }
+
+        timeTillNextPunchThroughAttempt_ -= dt;
+    }
+
+    // See if we need to still be trying to punch through from the client
+    if (clientConnectToGameServerState_ == GAME_CONNECTING_EXTERNAL_IP &&
+        clientToServerSocket_ != NULL && timeTillNextClientConnectAttempt_ <= 0)
+    {
+        Atomic::Network* network = GetSubsystem<Network>();
+
+        if (!network->GetServerConnection() || !network->GetServerConnection()->IsConnected())
+        {
+            LOGINFO("Sending packet to server");
```

Comment:
Would be good to remove this logging too, there are also log levels like LOGDEBUG and LOGERROR to use